File: templates/python/pp_plot.py
"""pp_plot: P-P 概率图（经验累积概率 vs 拟合正态理论累积概率）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from theme import apply_theme
from palette import cycle

def make_figure(title='P-P plot vs fitted normal'):
    apply_theme()
    rng = np.random.default_rng(9)
    n = 250
    samples = [
        ('normal sample', rng.normal(5, 2, n)),
        ('right-skewed sample', rng.gamma(2.0, 2.0, n)),
    ]
    fig, ax = plt.subplots(figsize=(5, 4.6))
    ax.plot([0, 1], [0, 1], '--', color='gray', linewidth=0.8, label='reference')
    p_emp = (np.arange(1, n + 1) - 0.5) / n
    for i, (name, x) in enumerate(samples):
        xs = np.sort(x)
        p_theo = stats.norm.cdf(xs, loc=x.mean(), scale=x.std(ddof=1))
        ax.scatter(p_theo, p_emp, s=12, color=cycle(i), alpha=0.6,
                   edgecolors='none', label=name)
    ax.set_xlim(-0.02, 1.02); ax.set_ylim(-0.02, 1.02)
    ax.set_xlabel('theoretical cumulative probability')
    ax.set_ylabel('empirical cumulative probability')
    ax.set_title(title)
    ax.legend(loc='upper left')
    ax.grid(True, linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/pp_plot.m
function fig = pp_plot()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(9);
    n = 250;
    x1 = 5 + 2*randn(n, 1);                              % normal sample
    x2 = -2*log(rand(n, 1)) - 2*log(rand(n, 1));         % gamma k=2 theta=2
    data = {x1, x2};
    names = {'normal sample', 'right-skewed sample'};
    fig = figure('Position',[100 100 500 460]); hold on;
    plot([0 1], [0 1], '--', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.8);
    p_emp = ((1:n)' - 0.5) / n;
    for i = 1:2
        x = data{i};
        xs = sort(x);
        mu = mean(x); sd = std(x);
        p_theo = 0.5 * (1 + erf((xs - mu) / (sd * sqrt(2))));  % fitted normal CDF
        scatter(p_theo, p_emp, 12, palette('cat',i), 'filled', ...
                'MarkerFaceAlpha', 0.6);
    end
    xlim([-0.02 1.02]); ylim([-0.02 1.02]);
    xlabel('theoretical cumulative probability');
    ylabel('empirical cumulative probability');
    title('P-P plot vs fitted normal');
    legend([{'reference'} names], 'Location', 'northwest');
    grid on;
end
